metricplugin/tcp.go

```diff
@@ -0,0 +1,347 @@
+package metricplugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-msgio"
+	"github.com/pkg/errors"
+)
+
+type tcpServer struct {
+	// The API of the parent plugin.
+	plugin PluginAPI
+
+	// A channel shared with all clients to notify them if we're shutting down.
+	closing chan struct{}
+	// A wait group to keep track of the clients.
+	wg sync.WaitGroup
+
+	// The underlying listener.
+	listener *net.TCPListener
+}
+
+// TCPServerConfig is the config for the metric export TCP server.
+type TCPServerConfig struct {
+	// ListenAddress specifies the address to listen on.
+	// It should take a form usable by `net.ResolveTCPAddr`, for example
+	// `localhost:8181` or `0.0.0.0:1444`.
+	ListenAddress string `json:"ListenAddress"`
+}
+
+func newTCPServer(plugin PluginAPI, cfg TCPServerConfig) (*tcpServer, error) {
+	addr, err := net.ResolveTCPAddr("tcp", cfg.ListenAddress)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to resolve listenAddress as TCP address")
+	}
+	listener, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to bind")
+	}
+	log.Infof("metric export TCP API running on %s", listener.Addr())
+
+	return &tcpServer{
+		plugin:   plugin,
+		listener: listener,
+		closing:  make(chan struct{}),
+		wg:       sync.WaitGroup{},
+	}, nil
+}
+
+func (s *tcpServer) listenAndServe() {
+	// See `Shutdown` about how this all works together.
+	defer func() {
+		log.Infof("cleaning up TCP clients")
+
+		close(s.closing)
+	}()
+
+	for {
+		// Accept a new connection.
+		c, err := s.listener.AcceptTCP()
+		if err != nil {
+			// We search for this error string to figure out if we're shutting
+			// down :)
+			if !strings.Contains(err.Error(), "use of closed network connection") {
+				log.Error("unable to accept connection: ", err)
+			}
+			return
+		}
+		log.Infof("new connection from %s", c.RemoteAddr())
+
+		remote := c.RemoteAddr()
+		client := tcpClient{
+			plugin:        s.plugin,
+			conn:          c,
+			remote:        remote,
+			writer:        msgio.NewWriter(c),
+			closed:        make(chan struct{}),
+			serverClosing: s.closing,
+			cleanup: func() {
+				s.wg.Done()
```

Comment:
This implicit variable passing is one reason why I hate go :)

Reply:
This has been cleaned up!